File: .planning/design/mockup-src/src/demo/store.ts
/**
 * Interactive-demo state store (design-review checkpoint feedback):
 * a plain reducer over dummy data seeded from `recipeFixtures.ts` — the
 * SAME single source of truth every static mockup route and the Go TUI
 * dummy render — so the live flows cannot drift from the approved copy.
 *
 * Everything here is dummy/in-memory: "writes" only mutate this state,
 * mirroring how the real product stages changes before its confirm +
 * backup ceremony. No backend, no persistence.
 */

import {
  healthFindings,
  identityManagerRows,
  type HealthFinding,
  type IdentityManagerRow,
  type IdentityManagerState,
  type MatchStrategy,
} from '../data/recipeFixtures';

export interface DemoIdentity extends IdentityManagerRow {
  /** Git author values, present once the identity's Git side is configured. */
  gitName?: string;
  gitEmail?: string;
  matchStrategy?: MatchStrategy;
  /** Real SSH endpoint + port (SSHUI-01); optional for seeded rows. */
  hostname?: string;
  port?: number;
}

/** STORE-01 dual storage strategy for the SSH config. */
export type SshStorageLayout = 'sentinel' | 'include';

export interface DemoFinding extends HealthFinding {
  /** Which identity this finding is about (drives per-identity health). */
  identity?: string;
}

export interface DemoState {
  identities: DemoIdentity[];
  findings: DemoFinding[];
  /** Whether the doctor scan has been run at least once this session. */
  scanned: boolean;
  /** global-ssh option keys the user applied via the fix ceremony. */
  sshApplied: string[];
  /** global-git baseline applied via the fix ceremony. */
  gitBaselineApplied: boolean;
  /** STORE-01: where gitid-managed SSH config lives. */
  sshStorage: SshStorageLayout;
  /** Timestamped backup paths "created" by write ceremonies, newest first. */
  backups: string[];
}

const seedIdentities: DemoIdentity[] = identityManagerRows.map((row) => ({
  ...row,
  ...(row.gitFragmentPath
    ? { gitName: `${row.name} identity`, gitEmail: `you@${row.name}.example` }
    : {}),
}));

const findingIdentity: Record<string, string> = {
  'ssh-key-perms-archived': 'archived',
  'ssh-identitiesonly-contradiction': 'clientB',
  'git-includeif-missing-fragment': 'legacy',
  'git-opensource-no-host-block': 'opensource',
};

const seedFindings: DemoFinding[] = healthFindings.map((f) => ({
  ...f,
  ...(findingIdentity[f.id] ? { identity: findingIdentity[f.id] } : {}),
}));

export const initialDemoState: DemoState = {
  identities: seedIdentities,
  findings: seedFindings,
  scanned: false,
  sshApplied: [],
  gitBaselineApplied: false,
  sshStorage: 'sentinel',
  backups: [],
};

export type DemoAction =
  | { type: 'add-identity'; identity: DemoIdentity; backup: string }
  | {
      type: 'configure-git';
      name: string;
      gitName: string;
      gitEmail: string;
      matchStrategy: MatchStrategy;
      backup: string;
    }
  | { type: 'clone-identity'; source: string; cloneName: string }
  | { type: 'delete-identity'; name: string; scope: 'everything' | 'git-only'; backup: string }
  | { type: 'new-key'; name: string; backup: string }
  | { type: 'mark-scanned' }
  | { type: 'fix-finding'; id: string; backup: string }
  | { type: 'apply-ssh'; keys: string[]; backup: string }
  | { type: 'apply-git-baseline'; backup: string }
  | {
      type: 'edit-ssh';
      name: string;
      sshHost: string;
      hostname: string;
      port: number;
      backup: string;
    }
  | { type: 'set-ssh-storage'; layout: SshStorageLayout; backup: string }
  | { type: 'reset' };

/** State an identity lands in once BOTH its SSH and Git sides exist. */
function recomputeAfterGit(row: DemoIdentity): IdentityManagerState {
  return row.sshHost ? 'complete' : 'git-only';
}

export function demoReducer(state: DemoState, action: DemoAction): DemoState {
  switch (action.type) {
    case 'add-identity':
      return {
        ...state,
        identities: [...state.identities, action.identity],
        backups: [action.backup, ...state.backups],
      };
    case 'configure-git':
      return {
        ...state,
        identities: state.identities.map((row) =>
          row.name === action.name
            ? {
                ...row,
                gitFragmentPath: `~/.gitconfig.d/${row.name}`,
                gitName: action.gitName,
                gitEmail: action.gitEmail,
                matchStrategy: action.matchStrategy,
                state: recomputeAfterGit(row),
                note: 'SSH Host block and Git fragment both present.',
              }
            : row,
        ),
        backups: [action.backup, ...state.backups],
      };
    case 'clone-identity': {
      const source = state.identities.find((row) => row.name === action.source);
      if (!source || state.identities.some((row) => row.name === action.cloneName)) {
        return state;
      }
      const clone: DemoIdentity = {
        ...source,
        name: action.cloneName,
        ...(source.sshHost ? { sshHost: `${action.cloneName}.github.com` } : {}),
        keyPath: `~/.ssh/id_ed25519_${action.cloneName}`,
        ...(source.gitFragmentPath
          ? { gitFragmentPath: `~/.gitconfig.d/${action.cloneName}` }
          : {}),
        note: `Cloned from "${action.source}" — new key + own Host block, same Git author.`,
      };
      return { ...state, identities: [...state.identities, clone] };
    }
    case 'delete-identity':
      if (action.scope === 'everything') {
        return {
          ...state,
          identities: state.identities.filter((row) => row.name !== action.name),
          findings: state.findings.filter((f) => f.identity !== action.name),
          backups: [action.backup, ...state.backups],
        };
      }
      return {
        ...state,
        identities: state.identities.map((row) => {
          if (row.name !== action.name) return row;
          const next: DemoIdentity = { ...row, state: 'incomplete' };
          delete next.gitFragmentPath;
          delete next.gitName;
          delete next.gitEmail;
          delete next.matchStrategy;
          next.note = 'SSH Host block present; Git identity was deleted.';
          return next;
        }),
        backups: [action.backup, ...state.backups],
      };
    case 'new-key':
      return {
        ...state,
        identities: state.identities.map((row) =>
          row.name === action.name
            ? {
                ...row,
                keyPath: `~/.ssh/id_ed25519_${row.name}`,
                ...(row.state === 'key-missing'
                  ? {
                      state: (row.gitFragmentPath
                        ? 'complete'
                        : 'incomplete') as IdentityManagerState,
                      note: 'New key generated; Host block re-points at it.',
                    }
                  : {}),
              }
            : row,
        ),
        backups: [action.backup, ...state.backups],
      };
    case 'mark-scanned':
      return { ...state, scanned: true };
    case 'fix-finding': {
      const finding = state.findings.find((f) => f.id === action.id);
      if (!finding) return state;
      let identities = state.identities;
      if (action.id === 'git-includeif-missing-fragment') {
        identities = identities.map((row) =>
          row.name === 'legacy'
            ? {
                ...row,
                state: 'complete',
                keyPath: row.keyPath ?? '~/.ssh/id_ed25519_legacy',
                note: 'Fragment restored — SSH Host block and Git fragment both present.',
              }
            : row,
        );
      }
      return {
        ...state,
        identities,
        findings: state.findings.filter((f) => f.id !== action.id),
        backups: [action.backup, ...state.backups],
      };
    }
    case 'apply-ssh':
      return {
        ...state,
        sshApplied: [...new Set([...state.sshApplied, ...action.keys])],
        backups: [action.backup, ...state.backups],
      };
    case 'apply-git-baseline':
      return {
        ...state,
        gitBaselineApplied: true,
        backups: [action.backup, ...state.backups],
      };
    case 'edit-ssh':
      return {
        ...state,
        identities: state.identities.map((row) =>
          row.name === action.name
            ? { ...row, sshHost: action.sshHost, hostname: action.hostname, port: action.port }
            : row,
        ),
        backups: [action.backup, ...state.backups],
      };
    case 'set-ssh-storage':
      return {
        ...state,
        sshStorage: action.layout,
        backups: [action.backup, ...state.backups],
      };
    case 'reset':
      return initialDemoState;
    default:
      return state;
  }
}

/** Header health rollup: worst live finding severity wins. */
export function healthRollup(state: DemoState): 'healthy' | 'warning' | 'error' {
  if (state.findings.some((f) => f.severity === 'error' || f.severity === 'critical')) {
    return 'error';
  }
  if (state.findings.some((f) => f.severity === 'warning')) return 'warning';
  return 'healthy';
}

/** Per-severity counts for the header chip (`N ids · ! w · ✗ e`). */
export function findingCounts(state: DemoState): { warnings: number; errors: number } {
  return {
    warnings: state.findings.filter((f) => f.severity === 'warning').length,
    errors: state.findings.filter((f) => f.severity === 'error' || f.severity === 'critical')
      .length,
  };
}

/** Fresh timestamped backup path in the same shape as the fixtures'. */
export function newBackupPath(file: string): string {
  const stamp = new Date()
    .toISOString()
    .replace(/:/g, '-')
    .replace(/\.\d+Z$/, 'Z');
  return `${file}.backup.${stamp}`;
}

export function findingsFor(state: DemoState, identityName: string): DemoFinding[] {
  return state.findings.filter((f) => f.identity === identityName);
}

File: internal/dummytui/store.go
package dummytui

// store.go is the Go mirror of
// .planning/design/mockup-src/src/demo/store.ts — a plain reducer over
// dummy data seeded from data.go (itself the Go mirror of
// recipeFixtures.ts, recipe-faithful per recipes/, the North Star).
//
// Everything here is dummy/in-memory: "writes" only mutate this state,
// mirroring how the real product stages changes before its confirm +
// backup ceremony. No backend, no persistence, no file I/O.

import (
	"strings"
	"time"
)

// DemoIdentity is one identity row of the live demo — an
// IdentityManagerRow extended with the Git author values, match strategy,
// and real SSH endpoint the web demo's DemoIdentity carries.
type DemoIdentity struct {
	Name            string
	State           string
	SSHHost         string
	KeyPath         string
	GitFragmentPath string
	Note            string
	// Git author values, present once the identity's Git side is configured.
	GitName       string
	GitEmail      string
	MatchStrategy string
	// Real SSH endpoint + port (SSHUI-01); optional for seeded rows.
	Hostname string
	Port     int
}

// SSHStorageLayout is STORE-01's dual storage strategy for the SSH config:
// "sentinel" (blocks in place inside ~/.ssh/config) or "include"
// (gitid-owned ~/.ssh/config.d/gitid.config via one Include line).
type SSHStorageLayout string

// The two STORE-01 storage layouts.
const (
	// StorageSentinel keeps gitid blocks sentinel-delimited in place.
	StorageSentinel SSHStorageLayout = "sentinel"
	// StorageInclude moves gitid blocks to an Include'd owned file.
	StorageInclude SSHStorageLayout = "include"
)

// DemoFinding is a HealthFinding attributed to the identity it is about
// (drives per-identity health); empty Identity means a global finding.
type DemoFinding struct {
	HealthFinding
	Identity string
}

// DemoState is the whole live-demo state — the Go mirror of store.ts's
// DemoState. Reduce transitions it; nothing else mutates it.
type DemoState struct {
	Identities []DemoIdentity
	Findings   []DemoFinding
	// Scanned is whether the doctor scan has run at least once this session.
	Scanned bool
	// SSHApplied holds global-ssh option keys applied via the ceremony.
	SSHApplied []string
	// GitBaselineApplied is whether the global-git baseline was applied.
	GitBaselineApplied bool
	// SSHStorage is STORE-01's current layout.
	SSHStorage SSHStorageLayout
	// Backups holds timestamped backup paths "created" by write
	// ceremonies, newest first.
	Backups []string
}

// findingIdentityAttribution maps seeded finding ids to the identity each
// finding is about — the Go mirror of store.ts's findingIdentity map.
// ssh-duplicate-host-star stays global (no entry).
var findingIdentityAttribution = map[string]string{
	"ssh-key-perms-archived":           "archived",
	"ssh-identitiesonly-contradiction": "clientB",
	"git-includeif-missing-fragment":   "legacy",
	"git-opensource-no-host-block":     "opensource",
}

// Seed builds the initial demo state from data.go's fixtures — the Go
// mirror of store.ts's initialDemoState. Rows with a Git fragment get the
// same derived author values the web seed uses.
func Seed() DemoState {
	identities := make([]DemoIdentity, 0, len(IdentityManagerRows))
	for _, row := range IdentityManagerRows {
		id := DemoIdentity{
			Name:            row.Name,
			State:           row.State,
			SSHHost:         row.SSHHost,
			KeyPath:         row.KeyPath,
			GitFragmentPath: row.GitFragmentPath,
			Note:            row.Note,
		}
		if row.GitFragmentPath != "" {
			id.GitName = row.Name + " identity"
			id.GitEmail = "you@" + row.Name + ".example"
		}
		identities = append(identities, id)
	}
	findings := make([]DemoFinding, 0, len(HealthFindings))
	for _, f := range HealthFindings {
		findings = append(findings, DemoFinding{
			HealthFinding: f,
			Identity:      findingIdentityAttribution[f.ID],
		})
	}
	return DemoState{
		Identities: identities,
		Findings:   findings,
		SSHStorage: StorageSentinel,
	}
}

// Action is one typed state transition — the Go mirror of store.ts's
// DemoAction union. Reduce is the only consumer.
type Action interface{ isAction() }

// AddIdentity appends a freshly-created identity (create wizard finish).
type AddIdentity struct {
	Identity DemoIdentity
	Backup   string
}

// ConfigureGit writes the Git side of an existing identity.
type ConfigureGit struct {
	Name          string
	GitName       string
	GitEmail      string
	MatchStrategy string
	Backup        string
}

// CloneIdentity copies Source as CloneName with its own new key path,
// Host alias, and fragment (MGR-04).
type CloneIdentity struct {
	Source    string
	CloneName string
}

// DeleteIdentity removes an identity — Scope "everything" drops the row
// and its findings; Scope "git-only" heals it to incomplete.
type DeleteIdentity struct {
	Name   string
	Scope  string // "everything" | "git-only"
	Backup string
}

// NewKey regenerates the identity's key (heals key-missing).
type NewKey struct {
	Name   string
	Backup string
}

// MarkScanned records that the doctor scan ran this session.
type MarkScanned struct{}

// FixFinding applies a doctor fix — the finding disappears and healed
// identities (legacy) change state.
type FixFinding struct {
	ID     string
	Backup string
}

// ApplySSH applies the chosen global-SSH option keys.
type ApplySSH struct {
	Keys   []string
	Backup string
}

// ApplyGitBaseline applies the global-git baseline managed block.
type ApplyGitBaseline struct {
	Backup string
}

// EditSSH rewrites an identity's managed Host block values.
type EditSSH struct {
	Name     string
	SSHHost  string
	Hostname string
	Port     int
	Backup   string
}

// SetSSHStorage migrates STORE-01's storage layout.
type SetSSHStorage struct {
	Layout SSHStorageLayout
	Backup string
}

// Reset restores the initial seeded state.
type Reset struct{}

func (AddIdentity) isAction()      {}
func (ConfigureGit) isAction()     {}
func (CloneIdentity) isAction()    {}
func (DeleteIdentity) isAction()   {}
func (NewKey) isAction()           {}
func (MarkScanned) isAction()      {}
func (FixFinding) isAction()       {}
func (ApplySSH) isAction()         {}
func (ApplyGitBaseline) isAction() {}
func (EditSSH) isAction()          {}
func (SetSSHStorage) isAction()    {}
func (Reset) isAction()            {}

// recomputeAfterGit is the state an identity lands in once BOTH its SSH
// and Git sides exist — the Go mirror of store.ts's recomputeAfterGit.
func recomputeAfterGit(row DemoIdentity) string {
	if row.SSHHost != "" {
		return "complete"
	}
	return "git-only"
}

// cloneState returns a deep copy of s so Reduce never mutates its input.
func cloneState(s DemoState) DemoState {
	next := s
	next.Identities = append([]DemoIdentity(nil), s.Identities...)
	next.Findings = append([]DemoFinding(nil), s.Findings...)
	next.SSHApplied = append([]string(nil), s.SSHApplied...)
	next.Backups = append([]string(nil), s.Backups...)
	return next
}

// Reduce applies action to state and returns the next state — a pure
// function mirroring store.ts's demoReducer transition-for-transition.
// The input state is never mutated.
func Reduce(state DemoState, action Action) DemoState { //nolint:gocyclo // one case per action type, mirroring store.ts's switch verbatim
	next := cloneState(state)
	switch a := action.(type) {
	case AddIdentity:
		next.Identities = append(next.Identities, a.Identity)
		next.Backups = append([]string{a.Backup}, next.Backups...)
	case ConfigureGit:
		for i, row := range next.Identities {
			if row.Name != a.Name {
				continue
			}
			row.GitFragmentPath = "~/.gitconfig.d/" + row.Name
			row.GitName = a.GitName
			row.GitEmail = a.GitEmail
			row.MatchStrategy = a.MatchStrategy
			row.State = recomputeAfterGit(row)
			row.Note = "SSH Host block and Git fragment both present."
			next.Identities[i] = row
		}
		next.Backups = append([]string{a.Backup}, next.Backups...)
	case CloneIdentity:
		var source *DemoIdentity
		for i := range next.Identities {
			if next.Identities[i].Name == a.Source {
				source = &next.Identities[i]
			}
			if next.Identities[i].Name == a.CloneName {
				return state // name taken — no-op, mirror store.ts
			}
		}
		if source == nil {
			return state
		}
		clone := *source
		clone.Name = a.CloneName
		if source.SSHHost != "" {
			clone.SSHHost = a.CloneName + ".github.com"
		}
		clone.KeyPath = "~/.ssh/id_ed25519_" + a.CloneName
		if source.GitFragmentPath != "" {
			clone.GitFragmentPath = "~/.gitconfig.d/" + a.CloneName
		}
		clone.Note = `Cloned from "` + a.Source + `" — new key + own Host block, same Git author.`
		next.Identities = append(next.Identities, clone)
	case DeleteIdentity:
		if a.Scope == "everything" {
			identities := next.Identities[:0]
			for _, row := range next.Identities {
				if row.Name != a.Name {
					identities = append(identities, row)
				}
			}
			next.Identities = identities
			findings := next.Findings[:0]
			for _, f := range next.Findings {
				if f.Identity != a.Name {
					findings = append(findings, f)
				}
			}
			next.Findings = findings
		} else {
			for i, row := range next.Identities {
				if row.Name != a.Name {
					continue
				}
				row.State = "incomplete"
				row.GitFragmentPath = ""
				row.GitName = ""
				row.GitEmail = ""
				row.MatchStrategy = ""
				row.Note = "SSH Host block present; Git identity was deleted."
				next.Identities[i] = row
			}
		}
		next.Backups = append([]string{a.Backup}, next.Backups...)
	case NewKey:
		for i, row := range next.Identities {
			if row.Name != a.Name {
				continue
			}
			row.KeyPath = "~/.ssh/id_ed25519_" + row.Name
			if row.State == "key-missing" {
				if row.GitFragmentPath != "" {
					row.State = "complete"
				} else {
					row.State = "incomplete"
				}
				row.Note = "New key generated; Host block re-points at it."
			}
			next.Identities[i] = row
		}
		next.Backups = append([]string{a.Backup}, next.Backups...)
	case MarkScanned:
		next.Scanned = true
	case FixFinding:
		found := false
		for _, f := range next.Findings {
			if f.ID == a.ID {
				found = true
			}
		}
		if !found {
			return state
		}
		if a.ID == "git-includeif-missing-fragment" {
			for i, row := range next.Identities {
				if row.Name != "legacy" {
					continue
				}
				row.State = "complete"
				if row.KeyPath == "" {
					row.KeyPath = "~/.ssh/id_ed25519_legacy"
				}
				row.Note = "Fragment restored — SSH Host block and Git fragment both present."
				next.Identities[i] = row
			}
		}
		findings := next.Findings[:0]
		for _, f := range next.Findings {
			if f.ID != a.ID {
				findings = append(findings, f)
			}
		}
		next.Findings = findings
		next.Backups = append([]string{a.Backup}, next.Backups...)
	case ApplySSH:
		for _, key := range a.Keys {
			present := false
			for _, existing := range next.SSHApplied {
				if existing == key {
					present = true
				}
			}
			if !present {
				next.SSHApplied = append(next.SSHApplied, key)
			}
		}
		next.Backups = append([]string{a.Backup}, next.Backups...)
	case ApplyGitBaseline:
		next.GitBaselineApplied = true
		next.Backups = append([]string{a.Backup}, next.Backups...)
	case EditSSH:
		for i, row := range next.Identities {
			if row.Name != a.Name {
				continue
			}
			row.SSHHost = a.SSHHost
			row.Hostname = a.Hostname
			row.Port = a.Port
			next.Identities[i] = row
		}
		next.Backups = append([]string{a.Backup}, next.Backups...)
	case SetSSHStorage:
		next.SSHStorage = a.Layout
		next.Backups = append([]string{a.Backup}, next.Backups...)
	case Reset:
		return Seed()
	}
	return next
}

// HealthRollup is the header health rollup: the worst live finding
// severity wins — "healthy", "warning", or "error".
func HealthRollup(state DemoState) string {
	rollup := "healthy"
	for _, f := range state.Findings {
		if f.Severity == SeverityError || f.Severity == SeverityCritical {
			return "error"
		}
		if f.Severity == SeverityWarning {
			rollup = "warning"
		}
	}
	return rollup
}

// FindingCounts holds the header chip's per-severity counts
// (`N ids · ! w · ✗ e`): errors folds error AND critical together.
type FindingCounts struct {
	Warnings int
	Errors   int
}

// CountFindings computes the live per-severity counts for the header chip.
func CountFindings(state DemoState) FindingCounts {
	var counts FindingCounts
	for _, f := range state.Findings {
		switch f.Severity {
		case SeverityWarning:
			counts.Warnings++
		case SeverityError, SeverityCritical:
			counts.Errors++
		case SeverityInfo:
		}
	}
	return counts
}

// FindingsFor returns the live findings attributed to identityName.
func FindingsFor(state DemoState, identityName string) []DemoFinding {
	var out []DemoFinding
	for _, f := range state.Findings {
		if f.Identity == identityName {
			out = append(out, f)
		}
	}
	return out
}

// NewBackupPath returns a fresh timestamped backup path in the same shape
// as the fixtures' (`<file>.backup.<ISO stamp with dashes for colons>`).
func NewBackupPath(file string) string {
	stamp := time.Now().UTC().Format("2006-01-02T15:04:05Z")
	return file + ".backup." + strings.ReplaceAll(stamp, ":", "-")
}
